[orc-rt] Add Error / Exception interop.

lhames opened the pull request:
The ORC runtime needs to work in diverse codebases, both with and without C++ exceptions enabled (e.g. most LLVM projects compile with exceptions turned off, but regular C++ codebases will typically have them turned on). This introduces a tension in the ORC runtime: If a C++ exception is thrown (e.g. by a client-supplied callback) it can't be ignored, but orc_rt::Error values will assert if not handled prior to destruction.  That makes the following pattern fundamentally unsafe in the ORC runtime:

```
if (auto Err = orc_rt_operation(...)) {
  log("failure, bailing out"); // <- may throw if exceptions enabled
  // Exception unwinds stack before Error is handled, triggers Error-not-checked
  // assertion here.
  return Err;
}
```

We can resolve this tension by preventing any exceptions from unwinding through ORC runtime stack frames. We can do this while preserving exception *values* by catching all exceptions (using `catch (...)`) and capturing their values as a std::exception_ptr into an Error.

This patch adds APIs to simplify conversion between C++ exceptions and Errors. These APIs are available only when enabled when the ORC runtime is configured with ORC_RT_ENABLE_EXCEPTIONS=On (the default).

- `ExceptionError` wraps a std::exception_ptr.

- `runCapturingExceptions` takes a T() callback and converts any exceptions thrown by the body into Errors. If T is Expected or Error already then runCapturingExceptions returns the same type. If T is void then runCapturingExceptions returns an Error (returning Error::success() if no exception is thrown). If T is any other type then runCapturingExceptions returns an Expected<T>.

- A new Error::throwOnFailure method is added that converts failing values into thrown exceptions according to the following rules:
1. If the Error is of type ExceptionError then std::rethrow_exception is called on the contained std::exception_ptr to rethrow the original exception value.
2. If the Error is of any other type then std::unique_ptr<T> is thrown where T is the dynamic type of the Error.

These rules allow exceptions to be propagated through the ORC runtime as Errors, and for ORC runtime errors to be converted to exceptions by clients.

lhames committed:
[orc-rt] Add Error / Exception interop.

The ORC runtime needs to work in diverse codebases, both with and without C++
exceptions enabled (e.g. most LLVM projects compile with exceptions turned off,
but regular C++ codebases will typically have them turned on). This introduces a
tension in the ORC runtime: If a C++ exception is thrown (e.g. by a
client-supplied callback) it can't be ignored, but orc_rt::Error values will
assert if not handled prior to destruction.  That makes the following pattern
fundamentally unsafe in the ORC runtime:

```
if (auto Err = orc_rt_operation(...)) {
  log("failure, bailing out"); // <- may throw if exceptions enabled
  // Exception unwinds stack before Error is handled, triggers Error-not-checked
  // assertion here.
  return Err;
}
```

We can resolve this tension by preventing any exceptions from unwinding through
ORC runtime stack frames. We can do this while preserving exception *values* by
catching all exceptions (using `catch (...)`) and capturing their values as a
std::exception_ptr into an Error.

This patch adds APIs to simplify conversion between C++ exceptions and Errors.
These APIs are available only when enabled when the ORC runtime is configured
with ORC_RT_ENABLE_EXCEPTIONS=On (the default).

- `ExceptionError` wraps a std::exception_ptr.

- `runCapturingExceptions` takes a T() callback and converts any exceptions
thrown by the body into Errors. If T is Expected or Error already then
runCapturingExceptions returns the same type. If T is void then
runCapturingExceptions returns an Error (returning Error::success() if no
exception is thrown). If T is any other type then runCapturingExceptions returns
an Expected<T>.

- A new Error::throwOnFailure method is added that converts failing values into
thrown exceptions according to the following rules:

  1. If the Error is of type ExceptionError then std::rethrow_exception is
     called on the contained std::exception_ptr to rethrow the original
     exception value.
  2. If the Error is of any other type then std::unique_ptr<T> is thrown where
     T is the dynamic type of the Error.

These rules allow exceptions to be propagated through the ORC runtime as Errors,
and for ORC runtime errors to be converted to exceptions by clients.

lhames commented:
@jaredwy @dwblaikie -- Flagging you guys, since you might be interested.

jaredwy commented on orc-rt/include/orc-rt/Error.h:
```diff
@@ -9,30 +9,79 @@
 
 #include <cassert>
 #include <cstdlib>
+#include <exception>
```
Yeah, wasn't sure if some places went to the extremes and removed it all together.

lhames committed:
Address review feedback:

1. Conditionalize `#include <exception>`: While this header should always be
safe to include we can save ourselves a little compile time by not parsing it
where it's not needed.

2. Simplify ExceptionError::toString and add a unit test for it.

lhames committed:
Address review feedback: Use make_unique.

lhames commented on orc-rt/include/orc-rt/Error.h:
```diff
@@ -127,10 +186,22 @@ class ORC_RT_NODISCARD Error {
+template <typename ThisT, typename ParentT>
+Error ErrorExtends<ThisT, ParentT>::restoreError() noexcept {
+  return make_error(
+      std::unique_ptr<ThisT>(new ThisT(*static_cast<ThisT *>(this))));
```
Hrm. I was thinking that we can't assume that `ThisT`'s constructor is available to `make_unique`, it's as unlikely (or more so) that it'd be available to ErrorExtends, so using make_unique seems reasonable.
